test/cross-chain/cross-chain.js

```diff
@@ -39,47 +58,153 @@ contract("Cross-chain", () => {
+
   before(async () => {
     await exec(`PORT=${FOREIGN_PORT} bash ./scripts/setup-foreign-chain.sh`);
 
     ({ bridgeMonitor, gnosisSafe, zodiacBridge, homeBridge, foreignBridge } = await setupBridging(homeRpcUrl, foreignRpcUrl));
 
-    // If Truffle is not on the home chain, then deploy colonyNetwork to the home chain
-    if (process.env.TRUFFLE_FOREIGN === "true") {
-      try {
-        await exec(`npm run provision:token:contracts`);
-        await exec(`npm run provision:safe:contracts`);
-        await exec(`npx truffle migrate --all --network development2`);
-      } catch (err) {
-        console.log(err);
-        process.exit();
-      }
+    getPromiseForNextBridgedTransaction = function () {
+      return new Promise((resolve, reject) => {
+        const listener = async function (_sender, msgSender, _messageId, success) {
+          console.log("bridged with ", _sender, msgSender, _messageId, success);
+          if (!success) {
+            console.log("bridged transaction did not succeed");
+            await reject(new Error("Bridged transaction did not succeed"));
+          }
+          homeBridge.off("RelayedMessage", listener);
+          foreignBridge.off("RelayedMessage", listener);
+          resolve();
+        };
+
+        homeBridge.on("RelayedMessage", listener);
+        foreignBridge.on("RelayedMessage", listener);
+      });
+    };
+
+    // Deploy colonyNetwork to whichever chain truffle hasn't already deployed to.
+    try {
+      await exec(`SOLIDITY_COVERAGE="" npx truffle compile --all`);
+      await exec(`SOLIDITY_COVERAGE="" npm run provision:token:contracts`);
+      await exec(`SOLIDITY_COVERAGE="" npm run provision:safe:contracts`);
+      await exec(`npx truffle migrate --network development2`);
+    } catch (err) {
+      console.log(err);
+
+      process.exit();
     }
-  });
 
-  beforeEach(async () => {
-    // Set up a colony on the home chain. That may or may not be the truffle chain...
-    // Get the etherrouter address
+    // Add bridge to the foreign colony network
     const homeNetworkId = await ethersHomeSigner.provider.send("net_version", []);
-    const homeChainId = await ethersHomeSigner.provider.send("eth_chainId", []);
+    // const homeChainId = await ethersHomeSigner.provider.send("eth_chainId", []);
+
+    const foreignNetworkId = await ethersForeignSigner.provider.send("net_version", []);
+    foreignChainId = await ethersForeignSigner.provider.send("eth_chainId", []);
+
     let etherRouterInfo;
     // 0x539 is the chain id used by truffle by default (regardless of networkid), and if
     // we see it in our tests that's the coverage chain, which builds the contract artifacts
     // in to a different location. If we see another chain id, we assume it's non-coverage
     // truffle and look for the build artifacts in the normal place.
-    if (homeChainId.toString() === "0x539") {
+    if (process.env.SOLIDITY_COVERAGE && process.env.TRUFFLE_FOREIGN === "false") {
```

Comment:
I would do a nested conditional here, a bit more clear IMO

Reply:
I don't think I understand the suggestion here. The best I think I can do with a nested conditional is:

```
    if (process.env.SOLIDITY_COVERAGE) {
      if (process.env.TRUFFLE_FOREIGN === "false"){
        etherRouterInfo = JSON.parse(fs.readFileSync("./build-coverage/contracts/EtherRouter.json"));
      } else {
        etherRouterInfo = JSON.parse(fs.readFileSync("./build/contracts/EtherRouter.json"));
      }
    } else {
      etherRouterInfo = JSON.parse(fs.readFileSync("./build/contracts/EtherRouter.json"));
    }
```

Which seems a lot worse to me?

Reply:
Bridging tests are run twice with coverage:

* Once with coverage tracking the chain where reputation mining is occuring (corresponding to `TRUFFLE_FOREIGN=false`)
* Once with coverage tracking the chain where reputation mining is not occurring (corresponding to `TRUFFLE_FOREIGN=true`)

Reply:
I don't think you've understood what's happening under coverage, here:

When these tests are run under coverage, there are two chains running. One is being run under coverage, and one is being run not under coverage. In the case of the former, the deployment artifacts are found in `./build-coverage/...`, and in the case of the latter the deployment artifacts are found in `./build/...`. That means under coverage, _both_ need to be loaded, but the question is which artifact corresponds to the mining-chain and which to the non-mining-chain, which `TRUFFLE_FOREIGN` captures.

I'll also note there's the case where neither run under coverage, which also needs to be accounted for in any solution here.